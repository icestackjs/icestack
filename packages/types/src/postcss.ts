export interface PrefixerOptions {
  prefix?: string
  ignore?: ReadonlyArray<RegExp | string>
}

export type VarPrefixerOptions = {
  varPrefix?: string // | PrefixFunction
  ignoreProp?: (RegExp | string)[]
  ignoreValueCustomProperty?: (RegExp | string)[]
}
