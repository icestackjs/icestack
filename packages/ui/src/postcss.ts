export * from '@icestack/postcss/ast'
export * from '@icestack/postcss/js'
export * from '@icestack/postcss/plugins'
export * from '@icestack/postcss/scss'
export { compressCssSelector, mapCss2JsArray, preprocessCssInJs, recursiveNodes, transformCss2Js } from '@icestack/postcss/utils'
export { AcceptedPlugin, CssInJs, PrefixerOptions, VarPrefixerOptions } from '@icestack/postcss/types'
