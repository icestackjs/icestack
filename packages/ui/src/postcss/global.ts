import type { PluginCreator, AcceptedPlugin, AtRule } from 'postcss'
import parser from 'postcss-selector-parser'
import type { CodegenOptions } from '@/types'
// media(hover:hover)

function isAtMediaHover(atRule: AtRule) {
  return /media\(\s*hover\s*:\s*hover\s*\)/.test(atRule.name) || (atRule.name === 'media' && /\(\s*hover\s*:\s*hover\s*\)/.test(atRule.params))
}
const creator: PluginCreator<CodegenOptions> = (options) => {
  const universal = options?.global?.selector?.universal
  const removeAtMediaHover = !options?.global?.atMedia?.hover
  const universalFn = typeof universal === 'string' ? () => universal : universal
  const ruleTransformer = parser((selectors) => {
    selectors.walk((selector) => {
      // universal
      if (selector.type === 'universal' && selector.value === '*') {
        const s = universalFn?.()
        if (s) {
          selector.value = s
        }
      }
    })
  })
  return {
    postcssPlugin: 'postcss-global-option-handler-plugin',
    plugins: [
      {
        postcssPlugin: 'deep-dark-fantastic',
        AtRule(atRule) {
          // hover

          if (removeAtMediaHover && isAtMediaHover(atRule)) {
            atRule.before(atRule.nodes)
            atRule.remove()
          }
        },
        Rule(rule) {
          ruleTransformer.transformSync(rule, { lossless: false, updateSelector: true })
        }
      }
    ] as AcceptedPlugin[]
  }
}

creator.postcss = true

export default creator
