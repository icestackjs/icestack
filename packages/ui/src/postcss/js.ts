import postcssJs from 'postcss-js'
import type { AcceptedPlugin } from 'postcss'
// import { getPlugin } from './custom-property-prefixer'
import postcssPrefix from './prefixer'
import globalPostcss from './global'
// import { rtlcss } from './rtlcss'
// import basePlugin from './runtime/base'
import { CodegenOptions } from '@/types'

export function getJsProcess(options: CodegenOptions) {
  const { runtime } = options
  const basePlugins: AcceptedPlugin[] = []
  const componentsPlugins: AcceptedPlugin[] = []
  const utilitiesPlugins: AcceptedPlugin[] = []

  // if (varPrefix) {
  //   const p = getPlugin(varPrefix)
  //   basePlugins.push(p)
  //   componentsPlugins.push(p)
  //   utilitiesPlugins.push(p)
  // }
  const { prefix } = runtime
  if (prefix) {
    const p = postcssPrefix(typeof prefix === 'string' ? { prefix, ignore: [] } : prefix)
    componentsPlugins.push(p)
    utilitiesPlugins.push(p)
  }
  // if (rtl) {
  //   const p = rtlcss(typeof rtl === 'boolean' ? undefined : rtl)
  //   componentsPlugins.push(p)
  //   utilitiesPlugins.push(p)
  // }

  if (runtime) {
    const p = globalPostcss(options)
    basePlugins.push(p)
    componentsPlugins.push(p)
    utilitiesPlugins.push(p)
  }

  // if (base) {
  //   const p = basePlugin(options)
  //   basePlugins.push(p)
  // }

  return {
    baseProcess: postcssJs.sync(basePlugins),
    componentsProcess: postcssJs.sync(componentsPlugins),
    utilitiesProcess: postcssJs.sync(utilitiesPlugins)
  }
}
