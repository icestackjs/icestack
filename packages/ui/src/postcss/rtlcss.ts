export { default as rtlcss } from 'rtlcss'
