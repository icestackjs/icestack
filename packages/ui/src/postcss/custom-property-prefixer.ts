import postcss from 'postcss'
import creator, { type UserDefinedOptions } from 'postcss-custom-property-prefixer'
import { defaultVarPrefix } from '@/constants'

export function getPlugin(varPrefix: UserDefinedOptions['prefix'] = defaultVarPrefix) {
  return creator({
    prefix: typeof varPrefix === 'string' ? varPrefix.slice(2) : varPrefix,
    ignoreProp: (decl) => {
      return decl.prop.startsWith('--tw-')
    },
    ignoreValueCustomProperty(customProperty) {
      return customProperty.startsWith('--tw-')
    }
  })
}

export async function addVarPrefix(rawCss: string, varPrefix: string = defaultVarPrefix) {
  const { css } = await postcss([getPlugin(varPrefix)])
    // @ts-ignore
    .process(rawCss, {
      from: undefined
    })
    .async()
  return css
}
