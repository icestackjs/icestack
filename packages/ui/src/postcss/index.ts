import postcssJs from 'postcss-js'
import postcss, { Root, AcceptedPlugin } from 'postcss'
import prefixer from './prefixer'
import type { CodegenOptions } from '@/types'
import type { Options as PrefixerOptions } from '@/postcss/prefixer'
export type { Options as PrefixerOptions } from '@/postcss/prefixer'
export { getPlugin as getCssVarsPrefixerPlugin } from './custom-property-prefixer'
export * from './tailwindcss'

export function resolvePrefixOption(options: string | PrefixerOptions): PrefixerOptions {
  return typeof options === 'string'
    ? {
        prefix: options
      }
    : options
}

export function getPrefixerPlugin(prefix: CodegenOptions['prefix']) {
  if (typeof prefix === 'string') {
    return prefixer({
      prefix
    })
  } else if (typeof prefix === 'object') {
    return prefixer(prefix)
  }
}

export function objectify(root: Root) {
  return postcssJs.objectify(root as Root)
}

export function process(plugins: AcceptedPlugin[], css: string) {
  // @ts-ignore
  return postcss(plugins).process(css, {
    from: undefined
  })
}

export { type Root, type AcceptedPlugin } from 'postcss'
