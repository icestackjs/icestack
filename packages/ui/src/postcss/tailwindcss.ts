import postcss from 'postcss'
import tailwindcss, { type Config } from 'tailwindcss'
import { CodegenOptions } from '..'
import gloablPostcss from './global'
export function resolveTailwindcss(opts: { css: string; config: Config; options: CodegenOptions }) {
  const { config, css, options } = opts

  // process.env.JEST_WORKER_ID = 'mock'
  // https://github.com/tailwindlabs/tailwindcss/blob/28e96baf3ac000f6ddca814a40ad402dbcf6a6a5/src/util/validateConfig.js#L7
  const result = postcss([tailwindcss(config), gloablPostcss(options)]) //
    // @tailwind base;\n
    // @ts-ignore
    // '@tailwind components;\n@tailwind utilities;\n' +
    .process(css, {
      from: undefined
    })
  // delete process.env.JEST_WORKER_ID
  return result
}

export function initConfig() {
  const config: Config = {
    content: [{ raw: 'hidden' }],
    theme: {
      extend: {}
    },
    corePlugins: {
      preflight: false
    }
  }
  return config
}
