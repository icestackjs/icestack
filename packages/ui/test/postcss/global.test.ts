import postcss from 'postcss'
import gloablPostcss from '@/postcss/global'
describe('global', () => {
  it('at media hover', () => {
    const { css } = postcss([gloablPostcss]).process(`@media (hover: hover) {
      &:hover {
        color:red;
      }
    }`)
    expect(css).toMatchSnapshot()
  })

  it('universal selector case 0', () => {
    const { css } = postcss([
      gloablPostcss({
        global: {
          selector: {
            universal: 'view'
          }
        }
      })
    ]).process(`.btm-nav > * {
      border-color: currentColor
    }`)
    expect(css).toMatchSnapshot()
  })

  it('pseudo root case 0', () => {
    const { css } = postcss([
      gloablPostcss({
        global: {
          selector: {
            root: 'page'
          }
        }
      })
    ]).process(`:root .countdown {
      line-height: 1em;
    }`)
    expect(css).toMatchSnapshot()
  })
})
