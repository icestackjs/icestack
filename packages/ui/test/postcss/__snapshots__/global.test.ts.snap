// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`global > at media hover 1`] = `
"
      &:hover {
        color:red;
      }"
`;

exports[`global > pseudo root case 0 1`] = `
"page .countdown {
      line-height: 1em;
    }"
`;

exports[`global > universal selector case 0 1`] = `
".btm-nav>view {
      border-color: currentColor
    }"
`;
